CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+/*
+ * This file was created by SET [Surgeon] [Hobbyist developer].
+ */
+#include "CrossSectionCompute.h"
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <math.h> // sqrt
+#include <vector>
+
+#include <vtkMRMLModelNode.h>
+#include <vtkMRMLMarkupsCurveNode.h>
+#include <vtkMRMLSegmentationNode.h>
+#include <vtkPlane.h>
+#include <vtkCutter.h>
+#include <vtkConnectivityFilter.h>
+#include <vtkContourTriangulator.h>
+#include <vtkMatrix4x4.h> // Should not be required.
+#include <vtkMassProperties.h>
+#include <vtkMath.h> // Pi()
+
+std::mutex mtx;
+
+#define WORKER_MESSAGE(msg) mtx.lock();\
+                            std::cout << msg << std::endl;\
+                            mtx.unlock();
+
+vtkStandardNewMacro(vtkCrossSectionCompute);
+
+void vtkCrossSectionCompute::PrintSelf(ostream& os, vtkIndent indent)
+{
+    vtkObject::PrintSelf(os,indent);
+    
+    os << indent << "numberOfThreads: " << this->numberOfThreads << "\n";
+    os << indent << "inputCenterlineNode: " << this->inputCenterlineNode << "\n";
+    os << indent << "inputSurfaceNode: " << this->inputSurfaceNode << "\n";
+    os << indent << "inputSegmentID: " << this->inputSegmentID << "\n";
+}
+
+// The surface polydata is constant. Create it once only.
+bool vtkCrossSectionCompute::SetInputSurfaceNode(vtkMRMLNode * inputSurface, const std::string& inputSegmentId)
+{
+  inputSurfaceNode = inputSurface;
+  inputSegmentID = inputSegmentId;
+  if (inputSurfaceNode->IsA("vtkMRMLSegmentationNode"))
+  {
+    vtkMRMLSegmentationNode * inputSegmentationNode = vtkMRMLSegmentationNode::SafeDownCast(inputSurfaceNode);
+    inputSegmentationNode->CreateClosedSurfaceRepresentation();
+    inputSegmentationNode->GetClosedSurfaceRepresentation(inputSegmentID, closedSurfacePolyData);
+  }
+  else if (inputSurfaceNode->IsA("vtkMRMLModelNode"))
+  {
+    vtkMRMLModelNode * inputModelNode = vtkMRMLModelNode::SafeDownCast(inputSurfaceNode);
+    closedSurfacePolyData->DeepCopy(inputModelNode->GetPolyData());
+  }
+  else
+  {
+    std::cout << "Invalid closed surface." << std::endl;
+    return false;
+  }
+  return true;
+}
+
+void vtkCrossSectionCompute::UpdateTable(vtkDoubleArray * crossSectionAreaArray, vtkDoubleArray * ceDiameterArray)
+{
+  if (inputCenterlineNode == NULL)
+  {     
+    std::cout << "Input centerline is NULL." << std::endl;
+    return;
+  }
+  if (inputSurfaceNode == NULL)
+  {     
+    std::cout << "Input surface is NULL." << std::endl;
+    return;
+  }
+  if (inputSegmentID.empty())
+  {     
+    std::cout << "Input segment ID is unknown." << std::endl;
+    return;
+  }
+  /*
+   * Divide the number of centerline points in equal blocks per thread.
+   * The last block will include the residual points also.
+   */
+  const unsigned int numberOfValues = crossSectionAreaArray->GetNumberOfValues();
+  unsigned int residual = numberOfValues % numberOfThreads;
```

Comment:
There is no need for messing with residual. Division operator or integer values will do integer division. result of `numberOfValues/numberOfThreads` is exactly the same as `(numberOfValues - residual) / numberOfThreads`.